Xact callbacks

mpalmi opened the pull request:
The set_user extension manipulates GUC state, and is therefore
responsible for maintaining it. In the event of transaction failure,
calls to the `(re)set_user` function were not correctly keeping track of
certain GUCs.

In an effort to make the extension handle transaction semantics more
robustly, introduce a transaction handler and add a formal state struct
to keep track of said state in a way that is (hopefully) clearer than
what was there before.

mpalmi committed:
Add set_user transaction callbacks

The set_user extension manipulates GUC state, and is therefore
responsible for maintaining it. In the event of transaction failure,
calls to the `(re)set_user` function were not correctly keeping track of
certain GUCs.

In an effort to make the extension handle transaction semantics more
robustly, introduce a transaction handler and add a formal state struct
to keep track of said state in a way that is (hopefully) clearer than
what was there before.

mpalmi committed:
Simplify reset_token logic and improve comments

gitstashpop commented on set_user.c:
```diff
@@ -60,12 +60,30 @@ PG_MODULE_MAGIC;
 static object_access_hook_type next_object_access_hook;
 
+/* set_user transaction state */
+static void set_user_xact_handler (XactEvent event, void *arg);
```
ws here for human readability would be good

gitstashpop commented on set_user.c:
```diff
@@ -60,12 +60,30 @@ PG_MODULE_MAGIC;
+} SetUserXactState;
+
+static SetUserXactState	*new;
+static SetUserXactState	*old;
```
new_state and old_state would be more clear

gitstashpop reviewed (commented):
lgtm

mpalmi committed:
Add transaction handler tests

mpalmi committed:
Add postgres version variants for transaction tests

There are some subtle differences in the outputs of the transaction handler
tests, due to varying GUC defaults and error outputs between postgres versions.
Account for this by making use of variant comparison files:
https://www.postgresql.org/docs/current/regress-variant.html

This commit specifically makes use of the second mechanism, where:
- set_user_1.out corresponds to variations in PostgreSQL version 9.6
- set_user_2.out corresponds to variations in PostgreSQL version 9.4
